packages/mcdoc/src/runtime/checker/index.ts

```diff
@@ -564,298 +705,303 @@ function condenseErrorsAndFilterSiblings<T>(
-		// non-simplified types and see if they have been compared yet. This is needed
-		// for structures that are cyclic, to essentially bail out once we are comparing
-		// the same types again and just collect the errors of the lower depth.
-		// This will currently lead to a stack overflow error when e.g. comparing two
-		// text component definitions
-		const simplifiedRuntime = simplify(inferredType, options, parents)
-		const inferredValue = getValueType(simplifiedRuntime)
-		const matches = []
-		for (const simplified of simplifiedOptions) {
-			const simplifiedValue = getValueType(simplified)
 
-			if (
-				(inferredValue.kind === simplifiedValue.kind ||
-					options.isEquivalent(inferredValue, simplifiedValue)) &&
-				(simplified.kind !== 'literal' ||
-					(simplifiedRuntime.kind === 'literal' &&
-						simplified.value.value === simplifiedRuntime.value.value)) &&
-				// TODO handle enum field attributes
-				(simplified.kind !== 'enum' ||
-					(simplifiedRuntime.kind === 'literal' &&
-						simplified.values.some(v =>
-							v.value === simplifiedRuntime.value.value
-						)))
-			) {
-				matches.push(simplified)
+	const childDefinitions: (McdocType | undefined)[] = Array(children.length)
+		.fill(undefined)
+	const errors: McdocCheckerError<T>[] = []
+	if (typeDef.attributes) {
+		for (const attribute of typeDef.attributes) {
+			const handler = attributeHandlers[attribute.name]
+			if (handler) {
+				errors.push(
+					...handler(
+						runtimeNode.originalNode,
+						attribute,
+						simplifiedInferred,
+						typeDef,
+						options,
+					),
+				)
 			}
 		}
+	}
+	switch (typeDef.kind) {
+		case 'any':
+		case 'unsafe':
+			break
+		case 'struct': {
+			const literalKvps = new Map<
+				string,
+				{
+					values: { pair: RuntimePair<T>; index: number }[]
+					definition: McdocType | undefined
+				}
+			>()
+			const otherKvps: { value: RuntimePair<T>; index: number }[] = []
 
-		if (matches.length === 0) {
-			possibleNode.validDefinitions = simplifiedOptions.map(d => ({
-				typeDef: d,
-				runtimeNode: possibleNode,
-				children: [],
-				errors: [{
-					kind: 'type_mismatch',
-					node: possibleNode.node,
-					expected: d,
-				}],
-			}))
-			continue
-		}
-		const children = options.getChildren(originalNode, simplifiedRuntime)
-
-		for (const simplified of matches) {
-			const errors: McdocCheckerError<T>[] = []
-			if (simplified.attributes) {
-				for (const attribute of simplified.attributes) {
-					const handler = attributeHandlers[attribute.name]
-					if (handler) {
-						errors.push(
-							...handler(
-								originalNode,
-								attribute,
-								simplifiedRuntime,
-								simplified,
-								options,
-							),
-						)
+			for (let i = 0; i < children.length; i++) {
+				const child = children[i]
+				if (Array.isArray(child)) {
+					continue
+				}
+				if (
+					child.key.inferredType.kind === 'literal' &&
+					child.key.inferredType.value.kind === 'string'
```

Comment:
Maybe it's time to add some helpers for these like `isStringLiteral(type: McdocType): type is { ... }`?